File: Day 22/solution.py
"""
Day 22 solution.
"""
import collections
import heapq
import itertools
import sys

X = Y = Z = int
ZYX = tuple[Z, Y, X]
"""Coordinates for the z, y, z axes respectively."""
XYZ = tuple[X, Y, Z]
"""Coordinates for the x, y, z axes respectively."""


class Solver:
    """Day 22 solver."""

    def __init__(self, filepath: str = "input.txt"):
        self.filepath = filepath

    def get_resting_bricks(
        self,
    ) -> tuple[list[tuple[XYZ, XYZ]], dict[XYZ, int]]:
        """Get the bricks after they are at rest and the blocks that the occupied by a brick."""

        def parse_coordinate(coordinate: str) -> ZYX:
            x, y, z = [int(x) for x in coordinate.split(",")]
            return z, y, x

        def parse_line(
            line: str,
        ) -> tuple[ZYX, ZYX]:
            start_coord, end_coord = (
                parse_coordinate(coordinate)
                for coordinate in line.strip().split("~")
            )
            lengths = tuple(e - s + 1 for s, e in zip(start_coord, end_coord))
            assert len(lengths) == 3
            return start_coord, lengths

        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            bricks = [parse_line(line) for line in file]

        # Find the resting positions for the bricks
        heapq.heapify(bricks)
        lowest_non_blocking_z: collections.defaultdict[
            tuple[X, Y], Z
        ] = collections.defaultdict(int)
        resting_bricks: list[tuple[XYZ, XYZ]] = []
        blocks_to_id: dict[XYZ, int] = {}
        while bricks:
            (z, y, x), lengths = heapq.heappop(bricks)
            l_z, l_y, l_x = lengths

            # Find the z position the brick will rest at
            max_blocking_z = 1 + max(
                lowest_non_blocking_z[x + d_x, y + d_y]
                for d_x, d_y in itertools.product(range(l_x), range(l_y))
            )
            min_z = min(max_blocking_z, z)

            # Mark the new highest z positions for each of the brick's xy positions
            for d_x, d_y in itertools.product(range(l_x), range(l_y)):
                n_x, n_y = (
                    x + d_x,
                    y + d_y,
                )
                lowest_non_blocking_z[n_x, n_y] = max(
                    lowest_non_blocking_z[n_x, n_y], min_z + l_z - 1
                )

            # Mark the blocks to belong to the current brick
            for d_z, d_y, d_x in itertools.product(
                *[range(l) for l in lengths]
            ):
                blocks_to_id[x + d_x, y + d_y, min_z + d_z] = len(
                    resting_bricks
                )

            resting_bricks.append(((x, y, min_z), (l_x, l_y, l_z)))

        return resting_bricks, blocks_to_id

    @staticmethod
    def create_directional_support_graphs(
        resting_bricks: list[tuple[XYZ, XYZ]], blocks_to_id: dict[XYZ, int]
    ) -> tuple[list[set[int]], list[set[int]]]:
        """
        Creates directional graphs mapping:
            - a brick to all the bricks it is supports
            - a brick to all the bricks it is supported by
        """
        supports = [set() for _ in range(len(resting_bricks))]
        supported_by = [set() for _ in range(len(resting_bricks))]
        for brick, ((x, y, z), (l_x, l_y, _)) in enumerate(resting_bricks):
            for d_x, d_y in itertools.product(range(l_x), range(l_y)):
                try:
                    brick_below = blocks_to_id[x + d_x, y + d_y, z - 1]
                except KeyError:
                    continue

                supported_by[brick].add(brick_below)
                supports[brick_below].add(brick)

        return supports, supported_by

    def part_1(self) -> int:
        """Part 1 solver."""
        resting_bricks, blocks_to_id = self.get_resting_bricks()

        supports, supported_by = self.create_directional_support_graphs(
            resting_bricks, blocks_to_id
        )

        return sum(
            all(
                len(supported_by[supported_brick]) > 1
                for supported_brick in supported_bricks
            )
            for supported_bricks in supports
        )

    def part_2(self) -> int:
        """Part 2 solver."""
        resting_bricks, blocks_to_id = self.get_resting_bricks()

        supports, supported_by = self.create_directional_support_graphs(
            resting_bricks, blocks_to_id
        )

        def count_falling_bricks(supported_bricks: set[int]) -> int:
            stack = [
                brick
                for brick in supported_bricks
                if len(supported_by[brick]) == 1
            ]

            falling_bricks = set(stack)
            while stack:
                brick = stack.pop()
                for candidate_brick in supports[brick]:
                    if candidate_brick in falling_bricks:
                        continue

                    if any(
                        support_brick not in falling_bricks
                        for support_brick in supported_by[candidate_brick]
                    ):
                        continue

                    falling_bricks.add(candidate_brick)
                    stack.append(candidate_brick)

            return len(falling_bricks)

        return sum(
            count_falling_bricks(supported_bricks)
            for supported_bricks in supports
        )

    def solve(self) -> None:
        """Runs part 1 and part 2."""
        print("Day 22")
        print(f"Solving: {self.filepath}")
        print("Part 1:")
        print(self.part_1())
        print("---")
        print("Part 2:")
        print(self.part_2())


if __name__ == "__main__":
    match sys.argv:
        case [_]:
            solver = Solver()
        case [_, "--path", path] | [_, "-p", path] | [_, path]:
            solver = Solver(path)
        case _:
            print("Usage: python solution.py <path>")
            sys.exit(0)
    solver.solve()

File: Day 22/solution.cpp
#include <filesystem>
#include <sstream>
#include <fstream>
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <stack>

// Gets the first item in a tuple.
template <typename T, typename... Ts>
auto head(std::tuple<T, Ts...> t)
{
	return std::get<0>(t);
}

// Creates a new tuple excluding the first item.
template <std::size_t... Ns, typename... Ts>
auto tail_impl(std::index_sequence<Ns...>, std::tuple<Ts...> t)
{
	return std::make_tuple(std::get<Ns + 1u>(t)...);
}

template <typename... Ts>
auto tail(std::tuple<Ts...> t)
{
	return tail_impl(std::make_index_sequence<sizeof...(Ts) - 1u>(), t);
}

// Recursive nested unordered map
template <typename V, typename... Ts>
class nested_unordered_map;

template <typename V, typename T>
class nested_unordered_map<std::tuple<T>, V>
{
	std::unordered_map<T, V> data;

public:
	V &operator[](const std::tuple<T> &key)
	{
		return this->data[head(key)];
	}

	V operator[](const std::tuple<T> &key) const
	{
		return this->data.at(head(key));
	}

	bool contains(const T &key) const
	{
		return this->data.contains(key);
	}
};

template <typename V, typename T, typename... Ts>
class nested_unordered_map<std::tuple<T, Ts...>, V>
{
	std::unordered_map<T, nested_unordered_map<std::tuple<Ts...>, V>> data;

public:
	V &operator[](const std::tuple<T, Ts...> &key)
	{
		return this->data[head(key)][tail(key)];
	}

	V operator[](const std::tuple<T, Ts...> &key) const
	{
		return this->data.at(head(key))[tail(key)];
	}

	bool contains(const T &key, const Ts &...rest) const
	{
		return this->data.at(key).contains(rest...);
	}
};

typedef std::tuple<int, int, int> ZYX;
typedef std::tuple<int, int, int> XYZ;

class Solver
{
	std::filesystem::path filepath;

	std::vector<std::pair<ZYX, ZYX>> get_bricks()
	{
		std::ifstream file(this->filepath);
		std::vector<std::pair<ZYX, ZYX>> bricks;
		for (char c = file.get(); c != EOF; c = file.get())
		{
			// Get start coordinates
			int start[3]{}, *current_num = &start[0];
			for (; c != '~'; c = file.get())
			{
				if (c == ',')
				{
					++current_num;
					continue;
				}

				*current_num = *current_num * 10 + c - '0';
			}

			// Get end coordinates
			int end[3]{};
			current_num = &end[0];
			for (char c = file.get(); c != '\n' && c != EOF; c = file.get())
			{
				if (c == ',')
				{
					++current_num;
					continue;
				}

				*current_num = *current_num * 10 + c - '0';
			}

			// Get length
			ZYX length({end[2] - start[2] + 1,
									end[1] - start[1] + 1,
									end[0] - start[0] + 1}),
					start_coord({start[2], start[1], start[0]});

			bricks.emplace_back(start_coord, length);
		}

		return bricks;
	}

	std::pair<std::vector<std::pair<XYZ, XYZ>>,
						nested_unordered_map<XYZ, int>>
	get_resting_bricks()
	{
		std::vector<std::pair<ZYX, ZYX>> bricks = this->get_bricks();

		auto heap_comp = std::greater<std::pair<ZYX, ZYX>>();
		std::make_heap(bricks.begin(), bricks.end(), heap_comp);
		nested_unordered_map<std::tuple<int, int>, int> lowest_non_blocking_z;
		std::vector<std::pair<XYZ, XYZ>> resting_bricks;
		nested_unordered_map<XYZ, int> blocks_to_id;
		while (!bricks.empty())
		{
			auto [coordinate, lengths] = bricks.front();
			std::pop_heap(bricks.begin(), bricks.end(), heap_comp);
			bricks.pop_back();
			auto [z, y, x] = coordinate;
			auto [l_z, l_y, l_x] = lengths;

			// Find the z position the brick will rest at
			int max_blocking_z = 0;
			for (int d_x = 0; d_x < l_x; ++d_x)
			{
				for (int d_y = 0; d_y < l_y; ++d_y)
				{
					max_blocking_z = std::max(max_blocking_z,
																		1 + lowest_non_blocking_z[{x + d_x, y + d_y}]);
				}
			}
			int min_z = std::min(max_blocking_z, z);

			// Mark the new highest z positions for each of the brick's xy positions
			for (int d_x = 0; d_x < l_x; ++d_x)
			{
				for (int d_y = 0; d_y < l_y; ++d_y)
				{
					std::tuple<int, int> position(x + d_x, y + d_y);
					lowest_non_blocking_z[position] = std::max(lowest_non_blocking_z[position],
																										 min_z + l_z - 1);
				}
			}

			// Mark the block to belong to the current brick
			for (int d_x = 0; d_x < l_x; ++d_x)
			{
				for (int d_y = 0; d_y < l_y; ++d_y)
				{
					for (int d_z = 0; d_z < l_z; ++d_z)
					{
						blocks_to_id[{x + d_x, y + d_y, min_z + d_z}] = resting_bricks.size();
					}
				}
			}
			resting_bricks.emplace_back(
					std::make_tuple(x, y, min_z),
					std::make_tuple(l_x, l_y, l_z));
		}

		return {resting_bricks, blocks_to_id};
	}

	static std::pair<
			std::vector<std::unordered_set<int>>,
			std::vector<std::unordered_set<int>>>
	create_directional_support_graphs(
			const std::vector<std::pair<XYZ, XYZ>> &resting_bricks,
			const nested_unordered_map<XYZ, int> &blocks_to_id)
	{
		std::vector<std::unordered_set<int>> supports(resting_bricks.size()),
				supported_by(resting_bricks.size());

		for (int brick = 0; brick < resting_bricks.size(); ++brick)
		{
			auto [x, y, z] = resting_bricks[brick].first;
			auto [l_x, l_y, _] = resting_bricks[brick].second;
			for (int d_x = 0; d_x < l_x; ++d_x)
			{
				for (int d_y = 0; d_y < l_y; ++d_y)
				{
					if (!blocks_to_id.contains(x + d_x, y + d_y, z - 1))
						continue;

					int brick_below = blocks_to_id[{x + d_x, y + d_y, z - 1}];
					supported_by[brick].insert(brick_below);
					supports[brick_below].insert(brick);
				}
			}
		}

		return {supports, supported_by};
	}

	int part_1()
	{
		auto [resting_bricks,
					blocks_to_id] = this->get_resting_bricks();

		auto [supports,
					supported_by] =
				this->create_directional_support_graphs(resting_bricks, blocks_to_id);

		int result = 0;
		for (const std::unordered_set<int> &supported_bricks : supports)
		{
			bool can_remove = true;
			for (int supported_brick : supported_bricks)
			{
				if (supported_by[supported_brick].size() <= 1)
				{
					can_remove = false;
					break;
				}
			}
			result += can_remove;
		}
		return result;
	}

	int part_2()
	{
		auto [resting_bricks,
					blocks_to_id] = this->get_resting_bricks();

		auto [supports,
					supported_by] =
				this->create_directional_support_graphs(resting_bricks, blocks_to_id);

		int result = 0;
		for (const std::unordered_set<int> &supported_bricks : supports)
		{
			std::stack<int> stack;
			std::unordered_set<int> falling_bricks;
			for (int brick : supported_bricks)
			{
				if (supported_by[brick].size() == 1)
				{
					stack.push(brick);
					falling_bricks.insert(brick);
				}
			}

			while (!stack.empty())
			{
				int brick = stack.top();
				stack.pop();
				for (int candidate_brick : supports[brick])
				{
					if (falling_bricks.contains(candidate_brick))
						continue;

					bool is_valid = true;
					for (int support_brick : supported_by[candidate_brick])
					{
						if (!falling_bricks.contains(support_brick))
						{
							is_valid = false;
							break;
						}
					}
					if (!is_valid)
						continue;

					falling_bricks.insert(candidate_brick);
					stack.push(candidate_brick);
				}
			}

			result += falling_bricks.size();
		}

		return result;
	}

public:
	Solver(std::filesystem::path filepath = "input.txt") : filepath(filepath) {}

	void solve()
	{
		std::cout << "Day 22" << std::endl;
		std::cout << "Solving: " << this->filepath << std::endl;
		std::cout << "Part 1:" << std::endl;
		std::cout << this->part_1() << std::endl;
		std::cout << "---" << std::endl;
		std::cout << "Part 2:" << std::endl;
		std::cout << this->part_2() << std::endl;
	}
};

int main(int argc, char *argv[])
{
	std::unique_ptr<Solver> solver;
	if (argc == 1)
	{
		solver = std::make_unique<Solver>();
	}
	else if (argc == 2)
	{
		solver = std::make_unique<Solver>(argv[1]);
	}
	else if (argc == 3 && (argv[1] == "-p" || argv[1] == "--path"))
	{
		solver = std::make_unique<Solver>(argv[2]);
	}
	else
	{
		return 1;
	}

	solver->solve();
}